atc/db/job_factory_test.go

```diff
@@ -730,52 +736,39 @@ var _ = Describe("Job Factory", func() {
 				It("fetches that job and the used resource", func() {
 					jobs, err := jobFactory.JobsToSchedule()
 					Expect(err).ToNot(HaveOccurred())
-					Expect(len(jobs)).To(Equal(3))
 
-					jobsMap := make(map[string]db.SchedulerJob)
 					for _, job := range jobs {
-						_, exists := jobsMap[job.Name()]
-						Expect(exists).To(BeFalse())
-
-						jobsMap[job.Name()] = job
+						switch job.Name() {
```

Comment:
I'm going to preface this by saying sorry for jumping around on this! Honestly, I think the test you had initially was mostly fine. I'm just curious to see if we can figure out a better way to do tests like these. Feel free to ignore me, though! Anyway:

I definitely feel this is better than a custom matcher, which would be almost certainly overkill. With that said, we do lose some coverage with this approach (at least the way it is now) - this test would pass in a few errant cases (there are no jobs; there are 3 jobs, but it's just "job1" repeated 3 times; etc). Now, we can add that coverage back by adding some additional assertions and logic here, but then we're basically back where we started!

I'm going to propose another possibility. Let me know what you think:

```go
...
// If we don't check this, we could have e.g. job1 present multiple times
Expect(len(jobs)).To(Equal(3))
jobResources := make(map[string][]db.SchedulerResource)
for _, job := range jobs {
    jobResources[job.Name()] = job.Resources
}
Expect(jobResources).To(ConsistOf(map[string]interface{}{
    job1.Name(): ConsistOf(
        db.SchedulerResource{
            Name: "some-resource",
            Type: "some-type",
        },
    ),
    job2.Name(): ConsistOf(...),
    job3.Name(): ConsistOf(...),
})
```

I haven't tested it, but it does appear `ConsistOf` can match on `maps`

Reply:
@aoldershaw Thats a really good point, using case statements can definitely give false positives! Haha don't apologize for that, its an iterative process and I'm really liking all the suggestions you are making. I think the approach you proposed is good, it might be unavoidable to be able to do this well without the weird mapping setup.

I was experimenting with using `MatchAllKeys` https://godoc.org/github.com/onsi/gomega/gstruct#MatchAllKeys rather than `ConsistOf` for comparing the map of scheduler resources? So it looks like this
```
Expect(jobResources).To(MatchAllKeys(Keys{
    job1.Name(): ConsistOf(
        db.SchedulerResource{
            Name: "some-resource",
            Type: "some-type",
        },
    ),
    job2.Name(): ConsistOf(...),
    job3.Name(): ConsistOf(...),
})
```
because I don't think the `ConsistOf` works with the nested `ConsistOf`s 🤔